### PR TITLE
NEW PLUGIN: Urql vue composition

### PR DESCRIPTION
**Pull request by Sytten**
## Description

This adds a new plugin to generator Vue composition functions for urql.

Related #6029

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## Screenshots/Sandbox (if appropriate/relevant):

Adding links to sandbox or providing screenshots can help us understand more about this PR and take action on it as appropriate

## How Has This Been Tested?

- [x] Unit tests
- [x] Code generation tests (typing works)

**Test Environment**:
- OS: macos
- `@graphql-codegen/...`: latest master
- NodeJS: 14

## Checklist:

- [x] I have followed the [CONTRIBUTING](https://github.com/the-guild-org/Stack/blob/master/CONTRIBUTING.md) doc and the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: packages/plugins/typescript/vue-urql/tests/__snapshots__/urql.spec.ts.snap [hunk 1]
```diff
@@ -0,0 +1,19 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
+
+exports[`urql Composition Should generate subscription hooks 1`] = `
+"import gql from 'graphql-tag';
```
Comment (BenoitRanque) [inline packages/plugins/typescript/vue-urql/tests/__snapshots__/urql.spec.ts.snap:4]: Is this double quote intentional? (not an authority on this, it just looks odd to me)

**Comment by Sytten** [comment]
@dotansimha when you have time to review :)
Unsure what I need to do to create the changeset, just let me know

**Comment by BenoitRanque** [comment]
I'm reviewing this at the moment, @Sytten if that's ok with you I'll make PRs to your branch.

About the changeset: I've been able to generate it, but I'm unclear what the versioning strategy should be.
For now, I'm setting the plugin version to `0.0.0`, with a changeset for a major version which will bring it back up to `1.0.0`
That said, the contribution documentation [here](https://www.graphql-code-generator.com/docs/custom-codegen/contributing#5-add-your-plugin) say that the plugin should be at the same version as the main plugin?
On the other hand not all packages share a version, so I am unsure what to do. @dotansimha if you could impart some wisdom it would be apreciated.

**Comment by BenoitRanque** [comment]
PR done, see [here](https://github.com/GloriousNurse/graphql-code-generator/pull/1) @Sytten 
Minor fixes, added changeset, added website documentation.
All tests pass.

Could not test the module itself in one of my projects, so my example in the docs is just my best guess.
I wanted to test the plugin itself in my project, but I could not despite my best efforts.
The tests pass so that will have to do I guess.

**Review by Sytten — COMMENTED** [review]
Seems ready for maintainer approval

**Comment by BenoitRanque** [comment]
Looking at the generated code, it seems we could remove the use of `any` and pass `never` instead?
(please take this with a grain of salt, I am very new to typescript)

this is example generated code:
```ts

```
as you can see we use `any` in the `UseQueryArgs` type. Here is the definition for that type
```ts
export interface UseQueryArgs<T = any, V = object> {
    query: MaybeRef<TypedDocumentNode<T, V> | DocumentNode | string>;
    variables?: MaybeRef<V>;
    requestPolicy?: MaybeRef<RequestPolicy>;
    context?: MaybeRef<Partial<OperationContext>>;
    pause?: MaybeRef<boolean>;
}
```
The generic `T` is only relevant if the query is of type `TypedDocumentNode<T, V>`.
Our generated code passes a `DocumentNode` instead, so the type `T` is not used.

Passing `never` would add extra type safety, and also silence some eslint errors 😉

**Note:** I am unsure how this plays with the `documentMode: external` and `importDocumentNodeExternallyFrom` options.
The code generated by `typescript-urql` does not use `any` at all

**Comment by Sytten** [comment]
I just passed `any` since I needed to put the variables in the second parameter, you can most likely pass a `never` or `unknown` there too instead. I don't have a strong opinion on this other that the default value for the parameter is `any` so that's what I put there.

**Comment by BenoitRanque** [comment]
Reason I'm suggesting this is mostly the ESLint rule [@typescript-eslint/no-explicit-any](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-explicit-any.md) which some projects may have enabled.

Since you don't have an objection, I will go ahead and change it to `never` which is more accurate, and less likely to trigger linting errors

**Comment by lukadriel7** [comment]
Hello, is this pull feature ready ? Is there anything left to do ?

**Comment by ardatan** [comment]
Sorry for late response.
Could you rebase the conflicts so we can run the tests?

**Comment by Sytten** [comment]
Will do today, if you can review the rest 🙏 @ardatan

**Comment by bartenra** [comment]
We are using:

```
  src/generated/graphql_urql.ts:
    plugins:
      - "typescript"
      - "typescript-operations"
      - "@graphql-codegen/typescript-urql"
    config:
      urqlImportFrom: "@urql/vue"
```

without any issues right now. 

The APIs could be different enough to warrant two different plugins, not sure about that.

File: packages/plugins/typescript/vue-urql/src/visitor.ts [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+export interface UrqlPluginConfig extends ClientSideBasePluginConfig {
+  withComposition: boolean;
+  urqlImportFrom: string;
+}
+
+export class UrqlVisitor extends ClientSideBaseVisitor<VueUrqlRawPluginConfig, UrqlPluginConfig> {
+  constructor(schema: GraphQLSchema, fragments: LoadedFragment[], rawConfig: VueUrqlRawPluginConfig) {
+    super(schema, fragments, rawConfig, {
+      withComposition: getConfigValue(rawConfig.withComposition, true),
+      urqlImportFrom: getConfigValue(rawConfig.urqlImportFrom, null),
```
Comment (bartenra) [inline packages/plugins/typescript/vue-urql/src/visitor.ts:22]: Why not set the default here?

File: website/docs/plugins/typescript-vue-urql.md [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+      }
+    }
+  }
+}
+```
+
+We can use the generated code like this:
+
+```ts
+  const [ result ] = useTestQuery(...);
```
Comment (bartenra) [inline website/docs/plugins/typescript-vue-urql.md:33]: `@urql/vue` doesn't return an array

**Comment by Sytten** [comment]
It does need another API because the typing is not exactly the same, I will apply your suggestions though.

**Comment by bartenra** [comment]
@Sytten I mean, wouldn't adding an option to the existing `typescript-urql` plugin be a possibility?

**Comment by Sytten** [comment]
On the contrary, it would just mean mixing react and vue code. I dont get why we would want to do that.

**Comment by hognevevle** [comment]
Anything I can do to help here?

**Comment by Sytten** [comment]
@ardatan @dotansimha I rebased and I think it is ready to merge.
@hognevevle I think we are all good.

**Comment by dotansimha** [comment]
Amazing work, thank you all :) 
We can move on with this one!

**Pull request merged**